src/kdc/kdc_authdata.c

```diff
@@ -792,8 +376,221 @@ get_auth_indicators(krb5_context context, krb5_enc_tkt_part *enc_tkt,
+ * result in a service ticket with a fresh PAC.  (3.3.5.7) says that
+ * otherwise, the PAC fields are copied from the TGT.  (Signatures are
+ * updated, though it does not mention this.)
+ *
+ * So if the incoming request had one (TGS), we verify it and attach it,
+ * resigned.  Otherwise, we ask the KDB for a fresh one.
+ */
+static krb5_error_code
+handle_pac(krb5_context context, unsigned int flags, unsigned int kdc_options,
+           krb5_db_entry *client, krb5_db_entry *server,
+           krb5_db_entry *subject_server, krb5_db_entry *local_tgt,
+           krb5_keyblock *local_tgt_key, krb5_keyblock *server_key,
+           krb5_const_principal altcprinc, krb5_enc_tkt_part *enc_tkt_reply,
+           krb5_ticket *old_ticket, krb5_ticket *new_ticket)
+{
+    krb5_error_code ret;
+    krb5_pac old_pac = NULL, new_pac = NULL;
+    krb5_boolean is_cross;
+    krb5_const_principal client_princ = NULL;
+    krb5_db_entry *krbtgt = local_tgt;
+    krb5_keyblock *krbtgt_key = local_tgt_key;
+
+    /* Select the right krbtgt for the realm. */
+    if (server_key != NULL)
+        krbtgt_key = server_key;
+    if (subject_server != NULL)
+        krbtgt = subject_server;
+
+    is_cross = is_cross_tgs_principal(krbtgt->princ);
+
+    old_pac = get_pac(context, enc_tkt_reply->authorization_data);
+    if (old_pac != NULL) {
+        if (old_ticket == NULL) {
+            /* Uh... TODO: wtf? */
+            ret = EINVAL;
+            goto done;
+        }
+
+        if (isflagset(flags, KRB5_KDB_FLAGS_S4U) && altcprinc != NULL)
+            client_princ = altcprinc;
+
+        /*
+         * The service key is known to the service, so there's no purpose
+         * checking it most of the time - just check the KDC signature.
+         * However, in a cross-realm TGS request, the KDC key is unknown to
+         * us, while the service key is the cross-realm key, so the two are
+         * "reversed".  Note that in this case, the Ticket Signature will also
+         * not be checkable.
+         */
+        if (is_cross) {
+            ret = krb5_kdc_verify_ticket(context, old_ticket, client_princ,
+                                         krbtgt_key, NULL);
+        } else {
+            ret = krb5_kdc_verify_ticket(context, old_ticket, client_princ,
+                                         NULL, krbtgt_key);
+        }
+        if (ret) {
+            /* While we could treat invalid PAC as nonexistent, it's simpler
+             * to just fail the request. */
+            goto done;
+        }
+    }
+    if (old_pac == NULL && (flags & NON_TGT_OPTION)) {
+        ret = 0;
+        goto done;
+    }
+
+    ret = minimal_pac(context, flags & KRB5_KDB_FLAG_CONSTRAINED_DELEGATION,
+                      client, server, old_pac, new_pac);
+    if (ret)
+        goto done;
+
+    /* Skip the KDB for modifications of tickets with valid PACs. */
+    if (old_pac == NULL || !(flags & NON_TGT_OPTION)) {
```

Comment:
For ticket modification requests, we can only get here we have a valid old PAC, so we only need the second part of this conditional.